pkg/gofr/datasource/file/ftp/file.go

```diff
@@ -0,0 +1,302 @@
+	}
+}
+
+// Close closes the FTP file connection.
+func (f *ftpFile) Close() error {
+	return f.response.Close()
+}
+
+// Name returns the name of the file.
+func (f *ftpFile) Name() string {
+	return f.name
+}
+
+// Read reads data from the FTP file into the provided byte slice and updates the file offset.
+func (f *ftpFile) Read(p []byte) (n int, err error) {
+	r, err := f.conn.RetrFrom(f.path, uint64(f.offset))
+	if err != nil {
+		return 0, err
+	}
+
+	b, err := io.ReadAll(r)
+	if err != nil {
+		return 0, err
+	}
+
+	r.Close()
+
+	copy(p, b)
+
+	readbytes := len(p)
+
+	if len(b) < len(p) {
+		f.offset += int64(len(b))
+		readbytes = len(b)
+
+		f.logger.Logf("Read %v bytes from %v", readbytes, f.path)
+
+		return len(b), io.EOF
+	}
+
+	f.logger.Logf("Read %v bytes from %v", readbytes, f.path)
+
+	f.offset += int64(readbytes)
+
+	return len(p), nil
+}
+
+// ReadAt reads data from the FTP file starting at the specified offset.
+func (f *ftpFile) ReadAt(p []byte, off int64) (n int, err error) {
+	resp, err := f.conn.RetrFrom(f.path, uint64(off))
+	if err != nil {
+		return 0, err
+	}
+
+	b, err := io.ReadAll(resp)
+	if err != nil {
+		return 0, err
+	}
+
+	copy(p, b)
+
+	resp.Close()
+
+	f.logger.Logf("Read %v bytes from %v at offset of %v", len(p), f.path, off)
+
+	return len(b), nil
+}
+
+func (f *ftpFile) Check(whence int, offset, length int64) (int64, error) {
+	switch whence {
+	case io.SeekStart:
+		if offset < 0 || offset > length {
+			return 0, ErrOutOfRange
+		}
+
+		return offset, nil
+
+	case io.SeekEnd:
+		if offset > 0 || offset < -length {
+			return 0, ErrOutOfRange
+		}
+
+		return length + offset, nil
+
+	case io.SeekCurrent:
+		if f.offset+offset >= length || f.offset+offset < 0 {
+			return 0, ErrOutOfRange
+		}
+
+		return f.offset + offset, nil
+	default:
+	}
+
+	return 0, ErrOutOfRange
+}
+
+// Seek sets the offset for the next Read or ReadAt operation.
+func (f *ftpFile) Seek(offset int64, whence int) (int64, error) {
+	var (
+		err error
+		r   ftpResponse
+	)
+
+	r, err = f.conn.Retr(f.path)
+	if err != nil {
+		return 0, err
+	}
+
+	p, err := io.ReadAll(r)
+	if err != nil {
+		return 0, err
+	}
+
+	err = r.Close()
+	if err != nil {
+		return 0, err
+	}
+
+	n := int64(len(p))
+
+	res, err := f.Check(whence, offset, n)
+
+	return res, err
+}
+
+// Write writes data to the FTP file.
+func (f *ftpFile) Write(p []byte) (n int, err error) {
+	reader := bytes.NewReader(p)
+
+	err = f.conn.StorFrom(f.path, reader, uint64(f.offset))
+	if err != nil {
+		return 0, errors.New("failed to write file")
+	}
+
+	f.offset += int64(len(p))
+
+	f.logger.Logf("Wrote %v bytes to %v", len(p), f.path)
+
+	return len(p), nil
+}
+
+// WriteAt writes data to the FTP file starting at the specified offset.
+func (f *ftpFile) WriteAt(p []byte, off int64) (n int, err error) {
+	reader := bytes.NewReader(p)
+
+	err = f.conn.StorFrom(f.path, reader, uint64(off))
+	if err != nil {
+		return 0, fmt.Errorf("failed to write at offset: %v", off)
```

Comment:
Same as above, you might want to wrap `err` so you know what the underlying error was?

Reply:
Noted.